mapping: allow $_dpmi=(0).

bartoldeman opened the pull request:
This is related to VCPI (#1880) but this should work no matter what. Allocate enough memory to allow for EXTMEM, XMS, and VGAEMU LFB mappings independent of DPMI settings.

Relatedly this cleans up config.vgaemu_memsize a bit so it doesn't need to be checked for 0 (which means 1MB for historical reasons) every time.

bartoldeman committed:
mapping: allow $_dpmi=(0).

This is related to VCPI (#1880) but this should work no matter what.
Allocate enough memory to allow for EXTMEM, XMS, and VGAEMU LFB mappings
independent of DPMI settings: if $_dpmi=(0), we reserve config.dpmi_base
memory (default: 32MB).

Relatedly this cleans up config.vgaemu_memsize a bit so it doesn't need to
be checked for 0 (which means 1MB for historical reasons) every time.

stsp commented on src/base/init/init.c:
```diff
@@ -344,6 +344,13 @@ void low_mem_init(void)
+  phys_rsv = EXTMEM_SIZE + config.xms_map_size;
+  /* High memory needs to be mapped for EXTMEM/XMS/VGAEMU independent
+     of config.dpmi and config.dpmi_lin_rsv_size values */
+  dpmi_size = _max(phys_rsv + config.vgaemu_memsize * 1024, dpmi_size);
```
Please use some other name, as
I guess this is not a dpmi_size now,
but some size to reserve.

stsp commented:
Lets make it slightly differently.
In dpmi_lin_mem_rsv() just remove
the check for config.dpmi. This is a
bug to keep it there as long as we
decided to abuse config.dpmi_lin_rsv_size
for the global main_pool.
Then in init.c just add the check that
config.dpmi_lin_rsv_size is large enough
for all the needs.

bartoldeman commented:
Yes the dpmi naming is confusing, on the other hand we don't need to reserve 512M when $_dpmi=(0).
Maybe just add to it after all like what's still in #1880, and don't include any overlaps in $_dpmi_lin_rsv_size? Then the default case would reserve (in total) 16MB+512MB+4MB.

For memsize and dpmi_size, I think it may be best to let memsize denote the total (present memsize+dpmi_size); the present memsize is just the constant LOWMEM_SIZE + HMASIZE.

bartoldeman commented:
Added a patch implementing this, it's just a result of adding, then search-replace memsize with LOWMEM_SIZE+HMASIZE, and then replacing LOWMEM_SIZE+HMASIZE+dpmi_size with memsize.

bartoldeman commented:
Note that now, if smalloc could be convinced to allocate at the top one could save one slot for KVM by putting the VGAEMU LFB all the way at the top (i.e. from 528GB to 532GB by default). Not sure how to do this best. That would just look cleaner, not make any difference in functionality though I think.

for example in
```
KVM: map slot=0 flags=0 dosaddr=0xff000000 size=0x00417000 unixaddr=0x7ff071800000
KVM: map slot=1 flags=0 dosaddr=0x00110000 size=0x0a4e6000 unixaddr=0x7ff04a510000
KVM: map slot=2 flags=0 dosaddr=0x00000000 size=0x000a0000 unixaddr=0x7ff04a200000
KVM: map slot=3 flags=0 dosaddr=0x000c0000 size=0x00050000 unixaddr=0x7ff04a2c0000
KVM: map slot=4 flags=1 dosaddr=0x000a0000 size=0x00020000 unixaddr=0x7ff04a2a0000
KVM: map slot=5 flags=0 dosaddr=0x0a9f6000 size=0x16a0a000 unixaddr=0x7ff054df6000
KVM: map slot=6 flags=1 dosaddr=0x0a5f6000 size=0x00400000 unixaddr=0x7ff0549f6000
```
slot 1 could then be merged with slot 5 and slot 6 would move to `dosaddr=0x21000000`

stsp commented:
There is smalloc_aligned_topdown().
Do you need a simple smalloc_topdown()?

stsp commented:
> Yes the dpmi naming is confusing, on the other hand we don't need to reserve 512M when $_dpmi=(0).

That's likely true, but not allocating
any reserve at all is what scares me.
Are you sure we won't smalloc()
something from a main_pool()?
I still suggest to put some logic
into dpmi_lin_mem_rsv(), and
probably set that param to
(auto) in the config?

bartoldeman commented:
aligned is ok. Looks like it could do the trick.

bartoldeman commented:
> That's likely true, but not allocating any reserve at all is what scares me. Are you sure we won't smalloc() something from a main_pool()? I still suggest to put some logic into dpmi_lin_mem_rsv(), and probably set that param to (auto) in the config?

Not sure I understand, now with `$_dpmi=(0)` it puts 20MB (16 phys + 4 vga) in `main_pool`, how does that mean "no reserve at all"?

stsp commented:
> Not sure I understand, now with $_dpmi=(0) it puts 20MB (16 phys + 4 vga) in main_pool,
> how does that mean "no reserve at all"?

But what is the reserve?
16Mb are eaten by extmem+xms, 4Mb
is eaten by lfb. And no reserve, smalloc()
will return OOM if someone tries to allocate.
512Mb - that's the reserve! :)
Of course w/o DPMI 512Mb is way too
much. Lets keep 32Mb reserve maybe?
Its allocated with PROT_NONE so does
not eat actual memory.

stsp commented:
So if DPMI is disabled, we can allocate
up to dpmi_base. As anything above
dpmi_base in in interest of dpmi only.
So what do you think about using dpmi_base
as a watermark?

bartoldeman commented:
yes dpmi_base as watermark looks ok to me.

stsp commented:
Then I guess the simplest logic would
be for dpmi_lin_mem_rsv() to return
dpmi_base + (config.dpmi ? dpmi_rsv : 0);
This will increase the default setup
by 32Mb over currently existing 512,
but this is probably not a problem, or
can be amended in a config. :)

bartoldeman commented:
That would actually use 1088k + dpmi_base then, so would need to be corrected to:
`dpmi_base - (LOWMEM_SIZE+HMASIZE) + (config.dpmi ? dpmi_rsv : 0);`
I found it easier to skip the subtraction followed by addition and just use init.c.

The vgaemu LFB placement (conditional on config.dpmi) is still a little awkward. It would actually be better to match a real PC with KVM, and place it at phys_addr=0xa0000000, then let the page tables take care of the linear (non-identity) mapping. That would also work with VCPI. For non-KVM dpmi=0 there is no need to map the LFB. But I think that should be for a different PR.

stsp commented:
So I guess this needs to be squashed.

bartoldeman commented:
ok, squashed!

stsp commented:
I didn't mean all 3, just 1 with 3. :)
But ok.
No, not OK, now your patch conflicts. :(
So you need to re-base, also can
reintroduce the separate clean-up
patch if you want.

bartoldeman committed:
mapping: memsize -> LOWMEM_SIZE+HMASIZE, memsize+dpmi_size -> memsize

memsize now denotes the total amount reserved at mem_base (main_pool.size),
instead of just the low portion, to have less confusing naming.

bartoldeman commented:
should be ok on the git side now

stsp commented on src/base/init/parser.y:
```diff
@@ -1096,7 +1096,7 @@ x_flag		: L_DISPLAY string_expr	{ free(config.X_display); config.X_display = $2;
 		| X_NOCLOSE bool      { config.X_noclose = ($2!=0); }
 		| X_NORESIZE bool     { config.X_noresize = ($2!=0); }
-		| VGAEMU_MEMSIZE expression	{ config.vgaemu_memsize = $2; }
+		| VGAEMU_MEMSIZE expression	{ config.vgaemu_memsize = ($2 ? $2 : 1024); }
```
Is this 1024 still needed?
I suppose we need to split lfb and
main vga mem?
Not in this patch-set of course.

stsp replied:
Converted this to config_scrub().

stsp commented:
Thanks!
